src/main/java/io/anserini/index/IndexUtils.java

```diff
@@ -183,6 +197,104 @@ public void printDocumentVector(String docid) throws IOException, NotStoredExcep
+    GzipCompressorOutputStream gzOut = new GzipCompressorOutputStream(bOut);
+    TarArchiveOutputStream tOut = new TarArchiveOutputStream(gzOut);
+
+    String sOut;
+    Map<Term, Integer> docFreqMap = new HashMap<>();
+    Map<String, String> docVectors;
+
+    int numNonEmptyDocs = reader.getDocCount(LuceneDocumentGenerator.FIELD_BODY);
+
+    String docid;
+    int counter = 0;
+    while ((docid = bRdr.readLine()) != null) {
+      counter++;
+
+      // get term frequency
+      Terms terms = reader.getTermVector(convertDocidToLuceneDocid(docid),
+              LuceneDocumentGenerator.FIELD_BODY);
+      if (terms == null) {
+        // We do not throw exception here because there are some
+        //  collections in which part of documents don't have document vectors
+        LOG.warn("Document vector not stored for doc " + docid);
+        continue;
+      }
+
+      TermsEnum te = terms.iterator();
+      if (te == null) {
+        LOG.warn("Document vector not stored for doc " + docid);
+        continue;
+      }
+
+      Term term;
+      long freq;
+
+      // iterate every term and write and store in Map
+      docVectors = new HashMap<>();
+      while ((te.next()) != null) {
+        term = new Term(LuceneDocumentGenerator.FIELD_BODY, te.term());
+        freq = te.totalTermFreq();
+
+        if (weight == null) {
+          // no weight
+          docVectors.put(term.bytes().utf8ToString(), String.valueOf(freq));
+        } else if (weight.equals("tf.idf")) {
+          // weighed by TFIDF
+          int docFreq;
+          if (docFreqMap.containsKey(term)) {
+            docFreq = docFreqMap.get(term);
+          } else {
+            try {
+              docFreq = reader.docFreq(term);
+            } catch (Exception e) {
+              LOG.error("Cannot find term " + term.toString() + " in indexing file.");
+              continue;
+            }
+            docFreqMap.put(term, docFreq);
+          }
+          float tfIdf =  (float) (freq * Math.log(numNonEmptyDocs * 1.0 / docFreq));
+          docVectors.put(term.bytes().utf8ToString(), String.format("%.6f", tfIdf));
+        } else {
+          // wrong weight argument
+          throw new InvalidDocVectorWeightException(weight + " is an invalid exception");
+        }
+      }
+
+      // Count size and write
+      int outLength = 0;
+      for (Map.Entry<String, String> entry: docVectors.entrySet()) {
+        outLength += (entry.getKey().getBytes(StandardCharsets.UTF_8).length
+                + entry.getValue().getBytes(StandardCharsets.UTF_8).length
+                + " \n".length());
+      }
+
+      TarArchiveEntry tarEntry = new TarArchiveEntry(new File(docid));
+      tarEntry.setSize(outLength + String.format("<DOCNO>%s</DOCNO>\n", docid).length());
+      tOut.putArchiveEntry(tarEntry);
+      tOut.write(String.format("<DOCNO>%s</DOCNO>\n", docid).getBytes());
+      for (Map.Entry<String, String> entry: docVectors.entrySet()) {
```

Comment:
Do you know about this?
```
Joiner.MapJoiner mapJoiner = Joiner.on(",").withKeyValueSeparator("=");
System.out.println(mapJoiner.join(map));
```
see: https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Joiner.MapJoiner.html

Reply:
How about 
```
String s = ranking.entrySet().stream().map(e -> e.getKey()+" "+e.getValue()).collect(joining("\n"));
```
?
Guava is less and less popular after Java 8 and Scala....

Reply:
Oh really? Should we kill it as a dependency then?

It was very much in use when I was at Twitter. Every method started with Guava precondition checks...

Reply:
I think we could file an issue for it.
In Java 8 and Scala/Rust, the introduction of `Option` forces the programmers to deal with `Null` which eliminates most of the precondition checks